### ISSUE TITLE
Surge Phenomenon in Optimal Learning Rate and Batch Size Scaling

### ISSUE THREAD

**Issue by Hiroki11x** [Hiroki11x/Papers#389]
## Links
https://arxiv.org/abs/2405.14578

## Authors
[Shuaipeng Li](https://arxiv.org/search/cs?searchtype=author&query=Li,+S), [Penghao Zhao](https://arxiv.org/search/cs?searchtype=author&query=Zhao,+P), [Hailin Zhang](https://arxiv.org/search/cs?searchtype=author&query=Zhang,+H), [Xingwu Sun](https://arxiv.org/search/cs?searchtype=author&query=Sun,+X), [Hao Wu](https://arxiv.org/search/cs?searchtype=author&query=Wu,+H), [Dian Jiao](https://arxiv.org/search/cs?searchtype=author&query=Jiao,+D), [Weiyan Wang](https://arxiv.org/search/cs?searchtype=author&query=Wang,+W), [Chengjun Liu](https://arxiv.org/search/cs?searchtype=author&query=Liu,+C), [Zheng Fang](https://arxiv.org/search/cs?searchtype=author&query=Fang,+Z), [Jinbao Xue](https://arxiv.org/search/cs?searchtype=author&query=Xue,+J), [Yangyu Tao](https://arxiv.org/search/cs?searchtype=author&query=Tao,+Y), [Bin Cui](https://arxiv.org/search/cs?searchtype=author&query=Cui,+B), [Di Wang](https://arxiv.org/search/cs?searchtype=author&query=Wang,+D)

## TLDR
Adam系オプティマイザでは、最適な学習率はバッチサイズに対して単調に増加するのではなく、まず上昇してピークに達した後に下降するという「サージ現象」を理論的に導出・実証した。このピーク時のバッチサイズは、学習速度とデータ効率のトレードオフ点に対応する。本研究は、大規模な深層学習モデルの効率的なハイパーパラメータ調整に貢献する。

最終的な Adam の critical batch size の推定 (B_noise (Adam 版)) は以下の通り。
これは計測できないので、fitting させて推定している（なのでオンラインで学習の途中で求めるのは難しい）

$$
B_{noise} = \frac{\pi \sum_{k} H_{k,k}}{2 \sum_{i, j, i \neq j} \frac{\mu_i \mu_j}{\sigma_i \sigma_j} H_{i,j}}
$$

## 1. はじめに：問題提起
深層学習において、Adam、RMSpropなどのAdam系オプティマイザは広く利用されています。これらのオプティマイザを効果的に機能させるには、学習率とバッチサイズの適切な設定が不可欠です。SGD（確率的勾配降下法）では、最適な学習率はバッチサイズに比例して増加する「線形スケーリング則」などが知られていますが、この法則はAdam系オプティマイザには当てはまりません。本論文は、Adam系オプティマイザにおける最適な学習率とバッチサイズの間の未知の関係性（スケーリング則）を解明することを目的としています。

## 2. 先行研究との違いと新規性
先行研究: Adam系オプティマイザについては、最適な学習率がバッチサイズの平方根(√B)に比例して単調に増加するという「平方根スケーリング」が提案されてきました。
新規性（本論文の主張）: 本論文は、この単調増加の関係はバッチサイズが小さい場合にしか当てはまらないと指摘します。代わりに、最適な学習率はバッチサイズの増加に伴い、まず上昇し、あるピーク（B_noiseと定義）に達した後に下降するという「サージ現象」が存在することを理論的に示しました。これは、Adam系が勾配の「符号（sign）」を用いて更新することに起因します。

## 3. 理論的貢献 (Theorems)
論文では、以下の仮定のもとで理論を展開しています。
- Adamのパラメータ更新は、勾配の符号 sign(Gest) で近似できる。
- ミニバッチの勾配 Gest は、中心極限定理によりガウス分布に従う。
- 損失関数は、局所的に二次関数で近似できる（テイラー展開）。

これらの仮定に基づき、以下の中心的なスケーリング則を導出しました。

$$
\epsilon_{opt}(B) = \frac{\epsilon_{max}}{\frac{1}{2}\left(\sqrt{\frac{B_{noise}}{B}} + \sqrt{\frac{B}{B_{noise}}}\right)}
$$

この式は、バッチサイズ B が B_noise と等しいときに、学習率 ε_opt が最大値 ε_max を取ることを示しています。BがB_noiseより小さい、あるいは大きい場合、学習率は低下します。
B_noise は、学習速度（少ないステップ数）とデータ効率（少ない学習サンプル数）のバランスが取れるトレードオフ点であり、学習が進んで損失が減少するにつれて、このB_noiseはより大きな値へとシフトしていくことも理論的に示しました。

## 4. 実証方法と結果
論文の理論的な主張を検証するため、多様なタスクで大規模な実験を行いました。

#### タスク:
Computer Vision (CV): CNN on Fashion-MNIST, ResNet-18 on Tiny-ImageNet
Natural Language Processing (NLP): DistilGPT2 on ELI5, MoE on RedPajama-v2

#### 方法: 
各タスクでバッチサイズと学習率を網羅的に探索するグリッドサーチを実施し、実際にどの組み合わせが最も性能が良いかを観測しました。

#### 結果: 
実験結果は、提案された「サージ現象」を明確に裏付けました。観測された最適な学習率の挙動は、先行研究の線形/平方根スケーリングよりも、本論文の理論曲線によくフィッティングしました。

## 5. 重要なFigureの説明
Figure 1: 本論文の核心的なアイデアを図示したものです。SGD（青線）では最適学習率が単調に増加して飽和するのに対し、Adam（オレンジ線）では一度ピークを迎えた後に下降する「サージ」が見られることを概念的に示しています。

<img width="674" height="365" alt="Image" src="https://github.com/user-attachments/assets/b3420513-2e17-463a-8fc8-e59210238685" />

Figure 3: ResNet-18の学習過程を示しています。学習が進む（損失L_trainingが減少する）につれて、最適学習率のピークを与えるバッチサイズ（赤い破線で示されたB_noise）が、より大きな値へと右に移動していることがわかります。これは、理論の予測と一致する重要な結果です。

<img width="735" height="238" alt="Image" src="https://github.com/user-attachments/assets/d1b2e305-5d22-437b-8267-39680fce3e51" />

<img width="657" height="515" alt="Image" src="https://github.com/user-attachments/assets/b7a4ca1e-5e17-4aed-994f-04701bbb65b1" />

Figure 5: 大規模なMoEモデルの実験結果です。本論文の提案する曲線（Ours, 緑色）が、観測された最適学習率（星印）を、線形スケーリング（Linear, 水色）や平方根スケーリング（Square root, 紫色）よりもはるかに正確に捉えていることを示しています。

<img width="542" height="382" alt="Image" src="https://github.com/user-attachments/assets/348b1c9f-7fc9-4469-9d0e-5b14805dd478" />

## この論文を引用している論文

- [Power Lines: Scaling Laws for Weight Decay and Batch Size in LLM Pre-training](https://arxiv.org/abs/2505.13738)
- [Small Batch Size Training for Language Models: When Vanilla SGD Works, and Why Gradient Accumulation Is Wasteful](https://arxiv.org/abs/2507.07101)
- [bs-scheduler: A Batch Size Scheduler library compatible with PyTorch DataLoaders](https://www.sciencedirect.com/science/article/pii/S2352711025001293)
- [Exploring Variance Reduction in Importance Sampling for Efficient DNN Training](https://arxiv.org/abs/2501.13296)
